rstar/src/primitives/geom_with_data.rs

```diff
@@ -0,0 +1,164 @@
+/// `geom` method in order to access geometry-specific methods.
+///
+/// # Examples
+/// ```
+/// use rstar::{RTree, PointDistance};
+/// use rstar::primitives::GeomWithData;
+///
+/// type RestaurantLocation = GeomWithData<&'static str, [f64; 2]>;
+///
+/// let mut restaurants = RTree::new();
+/// restaurants.insert(RestaurantLocation::new("Pete's Pizza Place", [0.3, 0.2]));
+/// restaurants.insert(RestaurantLocation::new("The Great Steak", [-0.8, 0.0]));
+/// restaurants.insert(RestaurantLocation::new("Fishy Fortune", [0.2, -0.2]));
+///
+/// let my_location = [0.0, 0.0];
+///
+/// // Now find the closest restaurant!
+/// let place = restaurants.nearest_neighbor(&my_location).unwrap();
+/// println!("Let's go to {}", place.data);
+/// println!("It's really close, only {} miles", place.distance_2(&my_location));
+/// ```
+/// ```
+/// // We can use any primitive type
+/// use rstar::{RTree, PointDistance};
+/// use rstar::primitives::GeomWithData;
+/// use rstar::primitives::Line;
+///
+/// let mut lines = RTree::new();
+/// lines.insert(GeomWithData::new(
+///     "Pete's Pizza Place",
+///     Line::new([0.0, 0.0], [1.0, 1.0]),
+/// ));
+/// lines.insert(GeomWithData::new(
+///     "The Great Steak",
+///     Line::new([0.0, 0.0], [-1.0, 1.0]),
+/// ));
+///
+/// let my_location = [0.0, 0.0];
+///
+/// // Now find the closest line
+/// let place = lines.nearest_neighbor(&my_location).unwrap();
+/// println!("Let's go to {}", place.data);
+/// println!(
+///     "It's really close, only {} miles",
+///     place.geom().distance_2(&my_location)
+/// );
+/// ```
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct GeomWithData<T, R> {
+    geom: R,
+    /// Data to be associated with the geometry being stored in the RTree
+    pub data: T,
+}
+
+impl<T, R: RTreeObject> RTreeObject for GeomWithData<T, R> {
+    type Envelope = R::Envelope;
+
+    fn envelope(&self) -> Self::Envelope {
+        self.geom.envelope()
+    }
+}
+
+impl<T, R: PointDistance> PointDistance for GeomWithData<T, R> {
+    fn distance_2(
+        &self,
+        point: &<Self::Envelope as Envelope>::Point,
+    ) -> <<Self::Envelope as Envelope>::Point as Point>::Scalar {
+        self.geom.distance_2(point)
+    }
+
+    fn contains_point(&self, p: &<Self::Envelope as Envelope>::Point) -> bool {
+        self.geom.contains_point(p)
+    }
+
+    fn distance_2_if_less_or_equal(
+        &self,
+        point: &<Self::Envelope as Envelope>::Point,
+        max_distance_2: <<Self::Envelope as Envelope>::Point as Point>::Scalar,
+    ) -> Option<<<Self::Envelope as Envelope>::Point as Point>::Scalar> {
+        self.geom.distance_2_if_less_or_equal(point, max_distance_2)
+    }
+}
+
+impl<T, R> GeomWithData<T, R> {
+    /// Creates a new [GeomWithData] with the provided data.
+    pub fn new(data: T, geom: R) -> Self {
```

Comment:
```suggestion
    pub fn new(geom: R, data: T) -> Self {
```

Would this be a more intuitive param ordering?

Geom-With-Data takes a Geom, then a Data?

Not a big deal, so take it or leave it.

If you do this, you might also want to flip the generic param order too, so it's all consistent (`GeomWithData<R, T>`)

Reply:
Yep, this is a thoughtful improvement.